.github/CODEOWNERS

```diff
@@ -0,0 +1,35 @@
+# Automatically request reviews from maintainers
+# The CODEOWNERS are listed alphabetically
+
+# Package
+/pybamm/discretisations/ @martinjrobins @tinosulzer
+/pybamm/experiment/ @brosaplanella @martinjrobins @tinosulzer @TomTranter
+/pybamm/expression_tree/ @martinjrobins @tinosulzer
+/pybamm/geometry/ @martinjrobins @tinosulzer
+/pybamm/input/ @brosaplanella @DrSOKane @tinosulzer @TomTranter
+/pybamm/meshes/ @martinjrobins @tinosulzer
+/pybamm/models/ @brosaplanella @DrSOKane @tinosulzer @TomTranter
+/pybamm/parameters/ @brosaplanella @DrSOKane @tinosulzer @TomTranter
+/pybamm/plotting/ @martinjrobins @Saransh-cpp @tinosulzer
+/pybamm/solvers/ @martinjrobins @tinosulzer @TomTranter
+/pybamm/spatial_methods/ @martinjrobins @tinosulzer
+/pybamm/step/ @brosaplanella @martinjrobins @tinosulzer @TomTranter
+/pybamm/* @pybamm-team/maintainers # the files directly under /pybamm/
+
+# CI/CD workflows
+/.github/ @martinjrobins @Saransh-cpp
+
+# Benchmarks
+/benchmarks/ @brosaplanella @Saransh-cpp
+
+# Documentation
+/docs/ @pybamm-team/maintainers
+
+# Example scripts
+/examples/ @pybamm-team/maintainers
+
+# Installation and other scripts
+/scripts/ @martinjrobins @Saransh-cpp
+
+# Files in the root directory
+/* @martinjrobins @Saransh-cpp
```

Comment:
I feel like this is a little too complicated and it is something that we would have to keep up-to-date.

I have not played around with overlapping rules before, so take this with a grain of salt. If we use scripts as an example:
```
# Installation and other scripts
/scripts/ @martinjrobins @Saransh-cpp
```
Does this mean only Saransh and Martin could review those files or does it let any maintainer approve and only they get notifications? If it is the former, then I think that would be a bit troublesome. The docs for code owners suggests that Saransh and Martin would be the owners for that directory, which sounds like the former case.

Reply:
Docs reference:
```
# These owners will be the default owners for everything in
# the repo. Unless a later match takes precedence,
# @global-owner1 and @global-owner2 will be requested for
# review when someone opens a pull request.
*       @global-owner1 @global-owner2

# Order is important; the last matching pattern takes the most
# precedence. When someone opens a pull request that only
# modifies JS files, only @js-owner and not the global
# owner(s) will be requested for a review.
*.js    @js-owner #This is an inline comment.
```

Reply:
I guess in my initial change I was suggesting that we turn on the require review from code owners in order to let trainees get more access. With this change as it is now, we get more notifications, but trainees still don't get more access.

Reply:
I think if we are not going to do the branch protection rules change, then we probably don't need to bother with a CODEOWNERS file.

Reply:
> Does this mean only Saransh and Martin could review those files or does it let any maintainer approve and only they get notifications? If it is the former, then I think that would be a bit troublesome. The docs for code owners suggests that Saransh and Martin would be the owners for that directory, which sounds like the former case.

I agree, we should aim for the latter.

For the parts of the code owned by pybamm-team/maintainers, does it mean that everyone will be requested to review?

Reply:
A single "maintainers" group will appear in the review section and each maintainer will get a single email. I believe that after somebody reviews the maintainers fer unassigned and the reviewer is the only one tagged. It has been a while since I used one of these files, but I believe that is how it works

Reply:
Does it auto-request a review or does it display an option for requesting reviews from the maintainers for those with the Triage level of permissions? It would be helpful to control the request behaviour.

Also, should we try testing this in someone's fork, since that would provide more clarity?